### PR TITLE
Dynamic protocol

### PR DESCRIPTION
**Pull request by jaxxzer**
update protocol fields and vector handling.

Needs https://github.com/bluerobotics/ping-protocol/pull/85

### CODE REVIEW

File: src/sensor/ping.cpp [hunk 1]
```diff
@@ -195,15 +198,17 @@ void Ping::handleMessage(PingMessage msg)
 
         // This is necessary to convert <uint8_t> to <int>
         // QProperty only supports vector<int>, otherwise, we could use memcpy
-        for (int i = 0; i < m.num_points(); i++) {
-            _points.replace(i, m.data()[i] / 255.0); // TODO we should really be working in ints
+        qCDebug(PING_PROTOCOL_PING) << "length:" << m.profile_data_length();
```
Comment (patrickelectric) [inline src/sensor/ping.cpp]: For high frequency messages, we should not use QCDebug. Right now all messages does have the same size.

Comment (jaxxzer) [reply src/sensor/ping.cpp]: whoops, should not be here.

File: src/link/pingsimulationlink.cpp [hunk 2]
```diff
@@ -21,16 +21,16 @@ void PingSimulationLink::randomUpdate()
...
     profile.set_distance(osc*(stop2+stop1)/(numPoints*2));
     profile.set_confidence(conf);
-    profile.set_pulse_usec(200);
+    profile.set_pulse_duration(200);
     profile.set_ping_number(counter);
     profile.set_scan_start(0);
     profile.set_scan_length(osc);
     profile.set_gain_index(4);
-    profile.set_num_points(200);
+    profile.set_profile_data_length(200);
```
Comment (patrickelectric) [inline src/link/pingsimulationlink.cpp]: numPoints

File: src/sensor/ping.cpp [hunk 3]
```diff
@@ -241,15 +243,17 @@ void Ping::handleMessage(PingMessage msg)
...
+
+        for (int i = 0; i < m.profile_data_length(); i++) {
+            _points.replace(i, m.profile_data()[i] / 255.0); // TODO we should really be working in ints
```
Comment (patrickelectric) [inline src/sensor/ping.cpp]: Why we should work with ints ?
The idea to use floats is to normalize the data from the sensor, 0 being the lowest and 1 the highest output that the sensor can return. This makes everything else after this point abstract for filters, widgets and etc.

File: src/sensor/ping.cpp [hunk 4]
```diff
@@ -503,13 +507,13 @@ void Ping::printStatus()
...
-    qCDebug(PING_PROTOCOL_PING) << "\t- pulse_usec:" << _pulse_usec;
+    qCDebug(PING_PROTOCOL_PING) << "\t- pulse_duration:" << _pulse_duration;
     qCDebug(PING_PROTOCOL_PING) << "\t- ping_number:" << _ping_number;
     qCDebug(PING_PROTOCOL_PING) << "\t- start_mm:" << _scan_start;
     qCDebug(PING_PROTOCOL_PING) << "\t- length_mm:" << _scan_length;
     qCDebug(PING_PROTOCOL_PING) << "\t- gain_index:" << _gain_index;
     qCDebug(PING_PROTOCOL_PING) << "\t- mode_auto:" << _mode_auto;
-    qCDebug(PING_PROTOCOL_PING) << "\t- msec_per_ping:" << _ping_rate;
+    qCDebug(PING_PROTOCOL_PING) << "\t- ping_interval:" << _ping_interval;
 //    qCDebug(PING_PROTOCOL_PING) << "\t- points:" << QByteArray((const char*)points, num_points).toHex();
```
Comment (patrickelectric) [inline src/sensor/ping.cpp]: num_points does not exist anymore.

**Comment by patrickelectric** [comment]
This PR fails to compile:
```
../lib/ping-protocol-cpp/src/protocol/pingmessage/pingmessage_ping1D.h:351:25: error: non-constant-expression cannot be narrowed from type 'int' to 'uint16_t' (aka 'unsigned short')
 in initializer list [-Wc++11-narrowing]
        : PingMessage { 36 + profile_data_length }
                        ^~~~~~~~~~~~~~~~~~~~~~~~
../lib/ping-protocol-cpp/src/protocol/pingmessage/pingmessage_ping1D.h:351:25: note: insert an explicit cast to silence this issue
        : PingMessage { 36 + profile_data_length }
                        ^~~~~~~~~~~~~~~~~~~~~~~~
                        static_cast<uint16_t>(  )

```

File: src/sensor/ping.cpp (hunk 3 above)
Comment (jaxxzer) [reply src/sensor/ping.cpp]: it's ok. Maybe I have been working without fpu too often.

would it be worth the effort/performance to normalize to 0~UINT16MAX?

File: src/link/pingsimulationlink.cpp (hunk 2 above)
Comment (jaxxzer) [reply src/link/pingsimulationlink.cpp]: ?

File: src/sensor/ping.cpp (hunk 3 above)
Comment (jaxxzer) [reply src/sensor/ping.cpp]: there is just a lot of high frequency math operations done against float from message->display that could be done in int math, that was my idea.

File: src/sensor/ping.cpp (hunk 4 above)
Comment (jaxxzer) [reply src/sensor/ping.cpp]: neither does `points`

**Comment by jaxxzer** [comment]
> This PR fails to compile:

ci passes

File: src/sensor/ping.cpp (hunk 1 above)
Comment (patrickelectric) [reply src/sensor/ping.cpp]: After doing some tests, I think that this message can be removed. This is almost half of the debug messages and does not help to debug, the profile has a fixed length, and if fails the checksum will show it to us.
```
[19:42:15:584] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:15:592] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:15:593] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:15:665] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:15:676] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:15:693] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:15:702] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:15:727] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:15:728] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:15:766] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:15:767] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:15:799] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:15:801] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:15:891] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:15:892] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:15:939] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:15:940] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:16:000] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:16:001] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:16:016] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:16:028] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:16:157] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:16:166] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:16:216] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:16:217] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:16:224] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:16:225] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:16:244] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:16:245] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:17:988] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:17:989] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:19:280] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:19:282] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:20:830] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:20:842] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:22:567] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:22:568] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:24:748] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:24:749] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:26:517] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:26:518] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:28:496] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:28:497] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:30:146] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:30:149] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:30:157] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:30:159] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:30:172] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:30:174] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:30:189] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:30:191] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:30:201] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:30:212] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:30:218] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:30:219] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:30:237] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:30:238] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:30:253] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:30:254] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:30:288] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:30:289] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
[19:42:30:296] ping.protocol.ping[Debug]: ping.cpp(149) Handling Message: 1300 Checksum Pass: true
[19:42:30:302] ping.protocol.ping[Debug]: ping.cpp(246) length: 200
```

File: src/sensor/ping.cpp [hunk 5]
```diff
@@ -503,14 +507,14 @@ void Ping::printStatus()
...
+    qCDebug(PING_PROTOCOL_PING) << "\t- pulse_duration:" << _pulse_duration;
     qCDebug(PING_PROTOCOL_PING) << "\t- ping_number:" << _ping_number;
     qCDebug(PING_PROTOCOL_PING) << "\t- start_mm:" << _scan_start;
     qCDebug(PING_PROTOCOL_PING) << "\t- length_mm:" << _scan_length;
     qCDebug(PING_PROTOCOL_PING) << "\t- gain_index:" << _gain_index;
     qCDebug(PING_PROTOCOL_PING) << "\t- mode_auto:" << _mode_auto;
-    qCDebug(PING_PROTOCOL_PING) << "\t- msec_per_ping:" << _ping_rate;
-//    qCDebug(PING_PROTOCOL_PING) << "\t- points:" << QByteArray((const char*)points, num_points).toHex();
+    qCDebug(PING_PROTOCOL_PING) << "\t- ping_interval:" << _ping_interval;
+    qCDebug(PING_PROTOCOL_PING) << "\t- points:" << QByteArray((const char*)_points.data(), _num_points).toHex();
```
Comment (patrickelectric) [inline src/sensor/ping.cpp]: You can do better prints with separator.
```
[20:05:30.091] ping.protocol.ping[Debug]: ping.cpp(502) Ping Status:
[20:05:30.093] ping.protocol.ping[Debug]: ping.cpp(503)         - srcId: 0
[20:05:30.094] ping.protocol.ping[Debug]: ping.cpp(504)         - dstID: 0
[20:05:30.094] ping.protocol.ping[Debug]: ping.cpp(505)         - device_type: 1
[20:05:30.094] ping.protocol.ping[Debug]: ping.cpp(506)         - device_model: 1
[20:05:30.095] ping.protocol.ping[Debug]: ping.cpp(507)         - fw_version_major: 3
[20:05:30.095] ping.protocol.ping[Debug]: ping.cpp(508)         - fw_version_minor: 22
[20:05:30.095] ping.protocol.ping[Debug]: ping.cpp(509)         - distance: 1480
[20:05:30.095] ping.protocol.ping[Debug]: ping.cpp(510)         - confidence: 63
[20:05:30.099] ping.protocol.ping[Debug]: ping.cpp(511)         - pulse_duration: 60
[20:05:30.100] ping.protocol.ping[Debug]: ping.cpp(512)         - ping_number: 50407
[20:05:30.100] ping.protocol.ping[Debug]: ping.cpp(513)         - start_mm: 0
[20:05:30.100] ping.protocol.ping[Debug]: ping.cpp(514)         - length_mm: 2263
[20:05:30.100] ping.protocol.ping[Debug]: ping.cpp(515)         - gain_index: 6
[20:05:30.100] ping.protocol.ping[Debug]: ping.cpp(516)         - mode_auto: true
[20:05:30.101] ping.protocol.ping[Debug]: ping.cpp(517)         - ping_interval: 0
[20:05:30.102] ping.protocol.ping[Debug]: ping.cpp(518)         - points: "c0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0
bfbfbfbfbfef3fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3fd1d0d0d0d0d0d03f101010101010903fc0bfbfbfbfbfef3fc0bfbfbfbfbfef3f"
```
E.g: 
```
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(502) Ping Status:
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(503)         - srcId: 0
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(504)         - dstID: 0
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(505)         - device_type: 1
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(506)         - device_model: 1
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(507)         - fw_version_major: 3
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(508)         - fw_version_minor: 22
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(509)         - distance: 792
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(510)         - confidence: 76
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(511)         - pulse_duration: 34
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(512)         - ping_number: 54359
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(513)         - start_mm: 0
[20:17:46.757] ping.protocol.ping[Debug]: ping.cpp(514)         - length_mm: 1157
[20:17:46.758] ping.protocol.ping[Debug]: ping.cpp(515)         - gain_index: 6
[20:17:46.758] ping.protocol.ping[Debug]: ping.cpp(516)         - mode_auto: true
[20:17:46.758] ping.protocol.ping[Debug]: ping.cpp(517)         - ping_interval: 0
[20:17:46.758] ping.protocol.ping[Debug]: ping.cpp(518)         - points: "c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f,c0,bf,bf,bf,bf,bf,ef,3f"  
```

Use `toHex(',');`

File: src/sensor/ping.cpp (hunk 1 above)
Comment (jaxxzer) [reply src/sensor/ping.cpp]: You are so right. I'm sorry to leave it here by mistake.

File: src/sensor/ping.h [hunk 6]
```diff
@@ -59,16 +59,16 @@ class Ping : public Sensor
      *
      * @return uint16_t firmware major version number
      */
-    uint16_t fw_version_major() { return _fw_version_major; }
-    Q_PROPERTY(int fw_version_major READ fw_version_major NOTIFY fwVersionMajorUpdate)
+    uint16_t firmware_version_major() { return _firmware_version_major; }
+    Q_PROPERTY(int firmware_version_major READ firmware_version_major NOTIFY firmwareVersionMajorUpdate)
```
Comment (patrickelectric) [inline src/sensor/ping.h]: All variables the are used in QML need to be updated to use the new name.

**Pull request merged**